Feature swc 67 añadir notificaciones por web socket para conexión y desconexión en lobbies

MatiasScantamburlo opened the pull request:
Notificar a jugadores  de un mismo lobby la conexión y desconexión de otro jugador **terminado** ¡Espero su revisión! 💯

MatiasScantamburlo committed:
feat: Implementa gestión de conexiones WebSocket por lobby

- Añadida la clase ManejadorConexionesLobby para manejar conexiones y
    desconexiones de jugadores en lobbies.
- Implementada la conexión y desconexión de jugadores a través del endpoint
    WebSocket /ws/lobby/{lobby_id}.
- Agregada funcionalidad de comunicación broadcast para notificar a todos los
    jugadores en el mismo lobby cuando un nuevo jugador se une o cuando un
    jugador abandona.

    Cambios a ser confirmados:
        nuevos archivos: actualizar_lobby.py

MatiasScantamburlo committed:
feat: Implementa gestión de conexiones WebSocket por lobby

- Añadida la clase ManejadorConexionesLobby para manejar conexiones y
    desconexiones de jugadores en lobbies.
- Implementada la conexión y desconexión de jugadores a través del endpoint
    WebSocket /ws/lobby/{lobby_id}.
- Agregada funcionalidad de comunicación broadcast para notificar a todos los
    jugadores en el mismo lobby cuando un nuevo jugador se une o cuando un
    jugador abandona.

    Cambios a ser confirmados:
        nuevos archivos: actualizar_lobby.py

MatiasScantamburlo committed:
test: Agrega test para notificaciones de conexión y desconexión de jugadores

- Implementado test_conectar_y_desconectar para verificar que los jugadores
    reciban notificaciones cuando un nuevo jugador se conecta o se desconecta.
- El test asegura que los mensajes de notificación se envíen correctamente a
    todos los jugadores conectados en el mismo lobby.

 Cambios a ser confirmados:
	modificados:     actualizar_lobby.py

MatiasScantamburlo committed:
test: Agrega test para notificaciones de conexión y desconexión de jugadores

- Implementado test_conectar_y_desconectar para verificar que los jugadores
    reciban notificaciones cuando un nuevo jugador se conecta o se desconecta.
- El test asegura que los mensajes de notificación se envíen correctamente a
    todos los jugadores conectados en el mismo lobby.

 Cambios a ser confirmados:
	modificados:     actualizar_lobby.py

MatiasScantamburlo committed:
test: Se agrega test para encapsulamiento de mensajeria en lobbies

- Implementado test_broadcast para verificar que los mensajes enviados a un
    lobby no se envíen a los jugadores en otros lobbies.
- Se implementa un manejador de timeout receive_text_with_timeout para
    gestionar el tiempo de espera en la recepción de mensajes.

 Cambios a ser confirmados:
	modificados:     actualizar_lobby.py

MatiasScantamburlo committed:
test: Se agrega test para encapsulamiento de mensajeria en lobbies

- Implementado test_broadcast para verificar que los mensajes enviados a un
    lobby no se envíen a los jugadores en otros lobbies.
- Se implementa un manejador de timeout receive_text_with_timeout para
    gestionar el tiempo de espera en la recepción de mensajes.

 Cambios a ser confirmados:
	modificados:     actualizar_lobby.py

MatiasScantamburlo committed:
Merge branch 'develop' into feature_SWC-67

- Tremos los cambios realizados en la db, posiblemente no hagan falta
  pero así ya vamos treabajando con los datos actuales y posiblemente
  evitemos conflictos.

MatiasScantamburlo committed:
Fix: Modularizacion del codigo

- Se agrega la carpeta services donde se almacena el manejador de
  conexiones al lobby
- Test se ubica en su archivo correspondiente

 Cambios a ser confirmados:
	modificados:     actualizar_lobby.py
	modificados:     main.py
	nuevos archivos: services/connection_manager.py
	nuevos archivos: test_actualizar_lobby.py

MatiasScantamburlo committed:
Fix: Modularizacion del codigo

- Se agrega la carpeta services donde se almacena el manejador de
  conexiones al lobby
- Test se ubica en su archivo correspondiente

 Cambios a ser confirmados:
	modificados:     actualizar_lobby.py
	modificados:     main.py
	nuevos archivos: services/connection_manager.py
	nuevos archivos: test_actualizar_lobby.py

MatiasScantamburlo committed:
Merge branch 'feature_SWC-67_Añadir_notificaciones_por_WebSocket_para_conexión_y_desconexión_en_lobbies' of github.com:HoodedHackers/Backend into feature_SWC-67_Añadir_notificaciones_por_WebSocket_para_conexión_y_desconexión_en_lobbies

MatiasScantamburlo committed:
fix: Modularizamos actualizar_lobby en main

- Pasamos las funcionalidades de actualizar_lobby dentro del main

 Cambios a ser confirmados:
	borrados:        actualizar_lobby.py
	modificados:     main.py
	renombrados:     test_actualizar_lobby.py -> tests/test_actualizar_lobby.py

KaiserOfNone commented:
No hagas el merge contra main, es contra develop

KaiserOfNone commented on main.py:
```diff
@@ -30,3 +33,20 @@ def get_games_repo(request: Request) -> GameRepository:
     return lobbies
+
+@app.websocket("/ws/lobby/{lobby_id}")
+async def websocket_endpoint(websocket: WebSocket, lobby_id: int):
```
Este nombre es demasiado generico. No me dice nada.

KaiserOfNone commented on main.py:
```diff
@@ -30,3 +33,20 @@ def get_games_repo(request: Request) -> GameRepository:
+@app.websocket("/ws/lobby/{lobby_id}")
+async def websocket_endpoint(websocket: WebSocket, lobby_id: int):
+    # Conectar al nuevo jugador a la lista de conexiones del lobby
+    await manejador.conectar(websocket, lobby_id)
```
1. Splanglish.
2. El `LobbyManager` o algo asi (no uses nombres genericos) deberia ser una dependencia onda como estamos haciendo los repositorios.

KaiserOfNone commented on main.py:
```diff
@@ -30,3 +33,20 @@ def get_games_repo(request: Request) -> GameRepository:
+    # Conectar al nuevo jugador a la lista de conexiones del lobby
+    await manejador.conectar(websocket, lobby_id)
+
+    await manejador.broadcast(f"Un nuevo jugador ha ingresado al lobby {lobby_id}.", lobby_id)
```
Porque mandas strings pelados. Acordamos que ibamos a mandar json

KaiserOfNone commented on main.py:
```diff
@@ -30,3 +33,20 @@ def get_games_repo(request: Request) -> GameRepository:
+    
+    try:
+        while True:
+            # Esto solo lo ponemos para mantener la conexión abierta
+            await websocket.receive_text()
```
Que estas esperando recibir. Esto no hace nada.

MatiasScantamburlo replied:
No espero nada,  solamente lo dejo para mantener la conexión abierta. Así entendí que funcionaba, sino hace falta y sin eso la conexión no se cierra, puedo sacarlo.

MatiasScantamburlo replied:
El comentario está desactualizado. En aquel momento creía que `receive_text` era necesario para mantener la conexión de ws abierta. Ahora esa linea está porque se espera recibir las notificaciones de entrada y salida de los demás jugadores a la sala

KaiserOfNone replied:
Pero estas leyendo y descartando. Esto literal no hace nada.

MatiasScantamburlo replied:
Ahí entendí Andy. Pensaba que había que capturar de alguna forma los mensajes, entre las conexiones, para que pudieran llegar al front, en mi cabeza esto lo hacía el `receive_text`. Ahí vi que de eso se encarga el `send_text` (o mejor dicho ahora lo hace el `send_json`, cuando suba los cambios). Así que habría que sacar la linea que me citaste :+1:

MatiasScantamburlo replied:
Sin embargo el while true debería tener algo que lo ponga al proceso en descanso para que no esté gastando CPU innecesariamente. Creo que en ese sentido es razonable dejar el `receive_text`.

KaiserOfNone replied:
Para saber que jugador es el que se esta conectando seguaremnte vas a tener que leer el identifier que te lo manden por el ws al inicio de la interaccion.

KaiserOfNone commented on main.py:
```diff
@@ -30,3 +33,20 @@ def get_games_repo(request: Request) -> GameRepository:
+        # Desconectar al jugador y notificamos a los demás jugadores
+        manejador.desconectar(websocket, lobby_id)
+
+        await manejador.broadcast(f"Un jugador ha abandonado el lobby {lobby_id}.", lobby_id)
```
Idem arriba. No mandes strings pelados.

KaiserOfNone commented on services/connection_manager.py:
```diff
@@ -0,0 +1,27 @@
+from typing import Dict, List
+from fastapi import WebSocket
+
+class ManejadorConexionesLobby:
```
Init? Conectar? Broadcast? Spanglish.

KaiserOfNone commented on services/connection_manager.py:
```diff
@@ -0,0 +1,27 @@
+        if lobby_id in self.lobbies:
+            self.lobbies[lobby_id].remove(websocket)
+            
+            if len(self.lobbies[lobby_id]) == 0:
```
No. El responsable de verificar si se debe borrar la conexion es el que llama
1. La funcion esta falla silenciosamente si le pedis borrar algo que no tiene.
2. Que pasa si queres borrar una partida aunque haya conexiones?

KaiserOfNone commented on tests/test_actualizar_lobby.py:
```diff
@@ -0,0 +1,54 @@
+from fastapi.testclient import TestClient
```
Igual que en el otro archivo. Estandariza a ingles. El resto de la base de codigo esta en ingles.

MatiasScantamburlo committed:
Refactoriza nombres, estandariza idioma y eviamos json

- Cambiados nombres genéricos por nombres más específicos para mejorar la
    claridad.
- Estandarizados los nombres de elementos de español a inglés en todo el
    código.
- Modificadas las funciones de broadcast para enviar datos en formato JSON en
    lugar de texto plano, mejorando la consistencia en la comunicación entre
    clientes y servidores.

 Cambios a ser confirmados:
	modificados:     main.py
	modificados:     services/connection_manager.py
	modificados:     tests/test_actualizar_lobby.py

MatiasScantamburlo committed:
Merge branch 'develop' into feature_SWC-67_Añadir_notificaciones_por_WebSocket_para_conexión_y_desconexión_en_lobbies

LourdesMamani00 commented on main.py:
```diff
@@ -1,11 +1,14 @@
 from sqlalchemy.orm import Session
 from fastapi import HTTPException
 from pydantic import BaseModel
 
```
tal vez debamos crear un archivo aparte para websockets pero creo que por el momeno si funciona en main entonces creo que para la entrega est'a

LourdesMamani00 replied:
parece que hay que actualizar la rama

MatiasScantamburlo committed:
Merge branch 'develop' into feature_SWC-67_Añadir_notificaciones_por_WebSocket_para_conexión_y_desconexión_en_lobbies

KaiserOfNone commented on main.py:
```diff
@@ -159,6 +158,23 @@ def get_games_available(repo: GameRepository = Depends(get_games_repo)):
+    await connection_manager.connect(websocket, lobby_id)
+
+    # Si se conecto un nuevo jugador al lobby notificamos a los demás jugadores
+    await connection_manager.broadcast(f"Un nuevo jugador se ha unido al lobby {lobby_id}.", lobby_id)
```
El manager es el que se encarga de que enviar me parece. Cuando lamas connect el mismo manager deberia avisar
Also envia algo como
```json
{ "player_name": "name" }
```

MatiasScantamburlo replied:
Estás bueno, avisar al front que la conexión ha sido exitosa enviando el nombre del jugador conectado

KaiserOfNone commented on main.py:
```diff
@@ -159,6 +158,23 @@ def get_games_available(repo: GameRepository = Depends(get_games_repo)):
+        # Desconectar al jugador y notificamos a los demás jugadores
+        connection_manager.disconnect(websocket, lobby_id)
+
+        await connection_manager.broadcast(f"Un jugador ha abandonado el lobby {lobby_id}.", lobby_id)
```
Mismo que el otro comment. Deja que el manager envie eso y manda un json.

MatiasScantamburlo replied:
El sistema de broadcast se encarga de enviar un JSON con el siguiente formato: {"message": <mensaje>} a un lobby específico.

Por ejemplo, en caso de una desconexión en el lobby 3, el endpoint correspondiente recibiría el siguiente mensaje:
{"message": "Un jugador ha abandonado el lobby 3"}

MatiasScantamburlo replied:
La razón por la cual no envió el mensaje directamente, a través del método connect o disconnect, es para generalizar la clase y mejorar su reúso en próximos endpoints ws

KaiserOfNone reviewed (commented):
`make format` is your friend :)

Mas que nada el tema es el diseño del manager, tipo el manager deberia ser el que se encarga de toda la logica de que se envia y el websocket_handler deberia solo levantar y bajar la conexion

MatiasScantamburlo committed:
Se agrega el método listen a la clase LobbyConnectionHandler y los test

- Inicia la conexión del WebSocket al lobby correspondiente y añade el socket
    a la lista de conexiones activas.
- Escucha mensajes en formato JSON de los clientes y verifica la acción
    solicitada (conectar, desconectar o acción inválida).
- Comprueba la existencia del jugador en el repositorio y que la acción
    solicitada sea válida.
- Captura excepciones de desconexión de WebSocket, asegurando que el sistema
    elimine correctamente el WebSocket de la lista de conexiones.
- Se agregan test que prueban la conexion y desconexion.

 Cambios a ser confirmados:
	modificados:     main.py
	modificados:     services/connection_manager.py
	borrados:        tests/test_actualizar_lobby.py
	nuevos archivos: tests/test_notify_lobby.py

KaiserOfNone commented on services/connection_manager.py:
```diff
@@ -1,27 +1,61 @@
+        await self.connect(websocket, lobby_id)
+        try:
+            while True:
+                data = await websocket.receive_json()
+                user_id = data.get("user_identifier")
+                player = player_repo.get_by_identifier(user_id)
+
+                if player is not None:
+                    user_name = player.name
+
+                    if data.get("action") == "connect":
+                        await self.broadcast(
+                            {"user_name": user_name, "action": "connect"}, lobby_id
+                        )
+
+                    elif data.get("action") == "disconnect":
+                        await self.broadcast(
+                            {"user_name": user_name, "action": "disconnect"}, lobby_id
+                        )
+                        await self.disconnect(websocket, lobby_id)
+
+                    else:
+                        await websocket.send_json({"error": "Invalid action"})
+                else:
+                    await websocket.send_json({"error": "User not found"})
```
Podes bajar el nivel de indentacion, mas que nada para evitar errores logicos, inverti los checkeos y hace un retorno especifico cuando sale algo mal. Tipo asi cuando alguien manda mal un json le llega un error raro poco descriptivo. Fijate con un test.

KaiserOfNone reviewed (changes requested):
El resto se ve bien, solo el comment que te dije y esta para mergear

MatiasScantamburlo committed:
Mejoramos la identacion y formato

LourdesMamani00 commented on services/connection_manager.py:
```diff
@@ -1,28 +1,64 @@
-    async def conectar(self, websocket: WebSocket, lobby_id: int):
-        # Añadimos la conexión del jugador a la lista lobby especificado
+    async def listen(
+        self, websocket: WebSocket, lobby_id: int, player_repo: PlayerRepository
```
creo que esta bien y como hablamos antes en la llamada supongo que vamos a ir modificando una que otra cosa a medida que avanzamos

MatiasScantamburlo committed:
colab Andy

KaiserOfNone committed:
SWC-67 Fix del test

KaiserOfNone committed:
SWC-67 Cambio en get_by_identifier para evitar accidentes

MatiasScantamburlo committed:
Avance en los test

- Se logra que un jugador notifique lista de jugadores por cada ingreso
- Falta comprobar notificaciones al egreso de un jugador
- Manejador se utiliza globalmente

 En la rama feature_SWC-67_Añadir_notificaciones_por_WebSocket_para_conexión_y_desconexión_en_lobbies
 Cambios a ser confirmados:
	modificados:     main.py
	modificados:     repositories/player.py
	modificados:     services/connection_manager.py
	modificados:     tests/test_notify_lobby.py

MatiasScantamburlo committed:
Cambiamos retorno de LobbyConnectionHandler.listen

- En vez de retornar el identificador se devuelve el id del jugador

 Cambios a ser confirmados:
	modificados:     services/connection_manager.py
	modificados:     tests/test_notify_lobby.py

MatiasScantamburlo committed:
Merge branch 'develop' into feature_SWC-67_Añadir_notificaciones_por_WebSocket_para_conexión_y_desconexión_en_lobbies

MatiasScantamburlo committed:
Merge branch 'develop' into feature_SWC-67_Añadir_notificaciones_por_WebSocket_para_conexión_y_desconexión_en_lobbies
